### PR TITLE
core_commands: use new show_room method etc

### PR DESCRIPTION
**Pull request by slook**
+ Changed: Adapt to recent core changes for calling `clear_message()`, `show_room()` and `remove_room()` methods
- Removed: No longer need to avoid joining already joined room as this scenario is now properly handled by the core
+ Added: Core check for joined room membership before attempting to leave room, but still allow offline tab closing
+ Added: Core check for joined room membership before sending any public chat message via plugin using `send_public()`
+ Changed: Core, when closing private chat, discard the user instead of remove, to avoid possible ValueError if not exist.

+ Added: `/ctcpversion` command
+ Added: Optional query argument for the `/help` command, enable finding specific commands, aliases or group names.

In theroey the commands should work from any interface, but they have only been tested in `chatrooms` and `private_chat`.

The different usages of the commands between public/private chats require some extra fiddling about with the optional/required `args` for the different `command_type`s. maybe built-in (user=None, room=None) might be more helpful than `source`, because then _trigger_command() could enforce the appropriate required arguments if the commands were duplicated with differing `usage`.

Ready for merging into your cli-commands-test branch @mathiascode , the "Users" and "Search" commands will be in separate PRs.

### CODE REVIEW

File: pynicotine/chatrooms.py [hunk 1]
```diff
@@ -69,6 +69,11 @@ def show_room(self, room, private=False):
             self.queue.append(slskmessages.JoinPublicRoom())
 
         elif room not in self.joined_rooms:
+            if room not in self.server_rooms and room not in self.private_rooms:
```
Comment (mathiascode) [inline pynicotine/chatrooms.py]: I don't think these log messages are needed right now.

File: pynicotine/chatrooms.py [hunk 2]
```diff
@@ -79,19 +84,18 @@ def remove_room(self, room):
...
+            self.queue.append(slskmessages.LeaveRoom(room))
+            self.joined_rooms.discard(room)
 
         if room in self.config.sections["columns"]["chat_room"]:
             del self.config.sections["columns"]["chat_room"][room]
 
         if self.ui_callback:
             self.ui_callback.remove_room(room)
 
-        self.core.pluginhandler.leave_chatroom_notification(room)
```
Comment (mathiascode) [inline pynicotine/chatrooms.py:93]: Why was this moved?

File: pynicotine/chatrooms.py [hunk 3]
```diff
@@ -187,6 +196,8 @@ def join_room(self, msg):
         if self.ui_callback:
             self.ui_callback.join_room(msg)
 
+        self.echo_message(msg.room, "Joined room %(room)s" % {"room": msg.room})
```
Comment (mathiascode) [inline pynicotine/chatrooms.py:194]: Debug?

File: pynicotine/chatrooms.py (hunk 2 above)
Comment (slook) [reply pynicotine/chatrooms.py:93]: because a notification was raised for leaving a mistyped room that didn't exist. The triggering of the notification before the room is actually left would seem to defeat the purpose of having this notification.

**Comment by mathiascode** [comment]
> The different usages of the commands between public/private chats require some extra fiddling about with the optional/required `args` for the different `command_type`s. maybe built-in (user=None, room=None) might be more helpful than `source`, because then _trigger_command() could enforce the appropriate required arguments if the commands were duplicated with differing `usage`.

Sounds like something we could try.

File: pynicotine/chatrooms.py (hunk 3 above)
Comment (slook) [reply pynicotine/chatrooms.py:194]: Yep. Although, I do quite like it. Some status is better than a blank screen, I consider that it could be useful for a console interface which doesn't have a widget for displaying the room name.

**Comment by slook** [comment]
The names of `source` and `command_type` are slightly ambiguous with each other (suggest `entity` and `interface`, respectively).

Alternatively, if we are going with the seperate `user` and `room` arguments, then plugin developers don't need to know the `command_type` and `source` anyway, because it will be obvious from checking which ones are set to `None`, i.e. `if room:`)

Then we might have `chatroom_usage` and `private_chat_usage` to include or exclude from the help and deal with args.

**Comment by mathiascode** [comment]
> Alternatively, if we are going with the seperate `user` and `room` arguments, then plugin developers don't need to know the `command_type` and `source` anyway, because it will be obvious from checking which ones are set to `None`, i.e. `if room:`)

This seems like the cleanest approach. I pushed a commit to your branch, let me know if it looks good to you.

File: pynicotine/chatrooms.py (hunk 2 above)
Comment (mathiascode) [reply pynicotine/chatrooms.py:93]: On second thought, I guess you get a join_chatroom_notification when reconnecting to the server, so sending one when closing a chatroom tab while disconnected doesn't make much sense.

> because a notification was raised for leaving a mistyped room that didn't exist.

We should bail out early so we don't trigger the UI callback.

Comment (slook) [reply pynicotine/chatrooms.py:93]: It is not really that clear whether `joined_rooms` is supposed to contain rooms that the user has selected to be in (i.e. open tabs), or if it is intended to be a list of active rooms that have been properly joined into by the server.

Since there seems to be some conflict about negotiating this between the client and server, perhaps this indicates the need for a seperate `opened_rooms` list that we always know reflects the open tabs in the client gui regardless of our ability or inability to establish a connection to the server for actually joining them.

That would help to ensure we don't accidentally leave any rooms by mistake that the user wanted to remain in, also help with allowing tabs to be closed while offline, and it may also potentially allow for us to build the ui widgets for displaying a status while waiting for the server to respond to the join request with the user list.

**Comment by slook** [comment]
It is definately heading in the right direction, that does seem more logical to implement this way.

Things were not working at first so I had to change the callbacks around, but I'm willing to change it again until we reach a satisfactory result. I don't know how you feel about the nested `"callbacks"` attribute in the command data, or whether it would be better to have them all on the top level (considering that we need to do something similar with the `"usage"` attributes)?

**Comment by mathiascode** [comment]
I don't quite understand, what was the issue with my approach?

I would prefer if we didn't have nested callbacks, they don't make much sense. If you need a different callback or usage message, you should add separate command entries to the self.chatroom_commands and self.private_chat_commands dicts.

**Comment by mathiascode** [comment]
> although it is frustrating that we can't seem to have both user and room in the same function together for dual purpose commands.

Just choose either user or room, depending on which one that isn't None. Or am I misunderstanding something?

**Comment by slook** [comment]
> I don't quite understand, what was the issue with my approach?

There was no issue, I was referring to the frustration as to my own comprehention of the best way for the code to be factored.

File: pynicotine/plugins/core_commands/__init__.py [hunk 4]
```diff
@@ -173,152 +247,118 @@ def help_command(self, _args, command_type, _source):
...
+        self.core.shares.rescan_shares()
+
+    def away_command(self, args=None, user=None, room=None):  # pylint: disable=unused-argument
         self.core.set_away_mode(self.core.user_status != 1, save_state=True)  # 1 = UserStatus.AWAY
-        self.echo_message("Status is now %s" % ("Online" if self.core.user_status == 2 else "Away"))
+        self.echo_message("Status is now %s" % (_("Online") if self.core.user_status == 2 else _("Away")))
 
-    def shutdown_notification(self):
-        self.log("Shutdown!")
+    def quit_program_chat_command(self, args=None, user=None, room=None):
```
Comment (slook) [inline pynicotine/plugins/core_commands/__init__.py]: [pylint:](https://github.com/mathiascode/nicotine-plus/actions/runs/3264462286/jobs/5365159352#step:6:10)

`pynicotine/plugins/core_commands/__init__.py:350:51: W0613: Unused argument 'user' (unused-argument)`

Using `_` to mark any unused keyword argument results in a TypeError.

File: pynicotine/pluginsystem.py [hunk 5]
```diff
@@ -729,25 +729,33 @@ def _trigger_command(self, command, args, user=None, room=None):
...
+                        plugin.echo_message("Usage: %s %s" % ('/' + command, " ".join(usage) if usage else ""))
+                        return
 
                     if room is not None:
-                        getattr(plugin, data.get("callback").__name__)(args, None, room)
+                        getattr(plugin, data.get("callback").__name__)(args, room=room)
 
                     elif user is not None:
-                        getattr(plugin, data.get("callback").__name__)(args, user, None)
+                        getattr(plugin, data.get("callback").__name__)(args, user=user)
```
Comment (slook) [inline pynicotine/pluginsystem.py:759]: TypeError is raised here if any unused keyword arguments are marked with _underscore, such as:

```
2022-10-17 11:19:04] Plugin core_commands failed with error <class 'TypeError'>: say_chat_command() got an unexpected keyword argument 'user'.
Trace:   File "/home/user/Git/slook/nicotine-plus/pynicotine/pluginsystem.py", line 758, in _trigger_command
    getattr(plugin, data.get("callback").__name__)(args, user=user)
```

Comment (mathiascode) [reply pynicotine/pluginsystem.py:759]: Ah, of course, I didn't test my changes after adding the underscores...

**Comment by mathiascode** [comment]
I tried to simplify the core commands plugin a bit, feel free to commit if it looks fine to you:

```python
# COPYRIGHT (C) 2022 Nicotine+ Team
#
# GNU GENERAL PUBLIC LICENSE
#    Version 3, 29 June 2007
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from pynicotine.pluginsystem import BasePlugin


class Plugin(BasePlugin):

    def __init__(self, *args, **kwargs):

        super().__init__(*args, **kwargs)

        commands = {
            "help": {
                "callback": self.help_command,
                "description": "Show commands",
                "usage": ["[query]"],
                "aliases": ["?"]
            },
            "rescan": {
                "callback": self.rescan_command,
                "description": _("Rescan shares"),
                "group": _("Shares")
            },
            "hello": {
                "callback": self.hello_command,
                "description": "Print something",
                "usage": ["[something..]"],
                "aliases": ["echo", "greet"],
                "group": _("Message")
            },
            "away": {
                "callback": self.away_command,
                "description": _("Toggle away status"),
                "aliases": ["a"]
            },
            "quit": {
                "callback": self.quit_command,
                "description": _("Quit Nicotine+"),
                "usage": ["[force]"],
                "choices": ["force"],
                "aliases": ["q", "exit"]
            }
        }

        chat_commands = {
            "clear": {
                "callback": self.clear_command,
                "description": _("Clear chat window"),
                "aliases": ["cl"],
                "choices": [-1],
                "group": _("Chat")
            },
            "join": {
                "callback": self.join_chat_command,
                "description": _("Join chat room"),
                "usage": ["<room>"],
                "aliases": ["j"],
                "group": _("Chat Rooms")
            },
            "me": {
                "callback": self.me_chat_command,
                "description": _("Say something in the third-person"),
                "usage": ["<something..>"],
                "group": _("Chat")
            },
            "msg": {
                "callback": self.msg_chat_command,
                "description": _("Send private message to user"),
                "usage": ["<user>", "<message..>"],
                "aliases": ["m"],
                "group": _("Private Chat")
            },
            "pm": {
                "callback": self.pm_chat_command,
                "description": _("Open private chat window for user"),
                "usage": ["<user>"],
                "group": _("Private Chat")
            },
            "say": {
                "callback": self.say_chat_command,
                "description": _("Say message in specified chat room"),
                "usage": ["<room>", "<message..>"],
                "group": _("Chat Rooms")
            },
            "ctcpversion": {
                "callback": self.ctcpversion_command,
                "description": _("Ask for a user's client version"),
                "usage": ["[user]"],
                "group": _("Client-To-Client Protocol")
            }
        }

        chatroom_commands = {
            "close": {
                "callback": self.close_command,
                "description": _("Close private chat"),
                "usage": ["<user>"],
                "aliases": ["c"],
                "group": _("Private Chat")
            },
            "leave": {
                "callback": self.leave_command,
                "description": _("Leave chat room"),
                "usage": ["[room]"],
                "aliases": ["l"],
                "group": _("Chat Rooms")
            },
        }

        private_chat_commands = {
            "close": {
                "callback": self.close_command,
                "description": _("Close private chat"),
                "usage": ["[user]"],
                "aliases": ["c"],
                "group": _("Private Chat")
            },
            "leave": {
                "callback": self.leave_command,
                "description": _("Leave chat room"),
                "usage": ["<room>"],
                "aliases": ["l"],
                "group": _("Chat Rooms")
            },
        }

        cli_commands = {
            "addshare": {
                "callback": self.add_share_command,
                "description": _("Add share"),
                "usage": ["<public|private>", "<virtual_name>", "<path>"],
                "group": _("Shares")
            },
            "removeshare": {
                "callback": self.remove_share_command,
                "description": _("Remove share"),
                "usage": ["<public|private>", "<virtual_name>"],
                "group": _("Shares")
            },
            "listshares": {
                "callback": self.list_shares_command,
                "description": _("List shares"),
                "group": _("Shares")
            }
        }

        self.chatroom_commands = {**commands, **chat_commands, **chatroom_commands}
        self.private_chat_commands = {**commands, **chat_commands, **private_chat_commands}
        self.cli_commands = {**commands, **cli_commands}

    def help_command(self, args, user=None, room=None):

        if user is not None:
            command_list = self.parent.private_chat_commands

        elif room is not None:
            command_list = self.parent.chatroom_commands

        else:
            command_list = self.parent.cli_commands

        query = args.split(" ", maxsplit=1)[0].lower().lstrip("/")
        command_groups = {}
        num_commands = 0

        for command, data in command_list.items():
            command_message = command
            usage = " ".join(data.get("usage", []))
            aliases = data.get("aliases", [])

            if aliases:
                command_message = command_message + " /" + " /".join(aliases)

            if usage:
                command_message += " " + usage

            description = data.get("description", "No description")
            group = data.get("group", _("Commands"))
            group_words = group.lower().split(" ")

            if not args or query in command or query in (a for a in aliases) or query in group_words:
                if group not in command_groups:
                    command_groups[group] = []

                command_groups[group].append("    %s  -  %s" % (command_message, description))
                num_commands += 1

        if not num_commands:
            self.echo_unknown_command(query)

        elif num_commands >= 2 and query:
            self.echo_message("List of %i commands matching \"%s\":" % (num_commands, query))

        for group, commands in command_groups.items():
            self.echo_message("")
            self.echo_message("  " + group + ":")

            for command in commands:
                self.echo_message(command)

    def clear_command(self, args, user=None, room=None):

        if room is not None:
            self.core.chatrooms.clear_messages(room)

        elif user is not None:
            self.core.privatechats.clear_messages(user)

    def close_command(self, args, user=None, **_unused):

        if args:
            user = args

        if user in self.core.privatechats.users:
            self.echo_message("Closing private chat of user %s" % user)
        elif user:
            self.echo_message("Not messaging with user %s" % user)

        self.core.privatechats.remove_user(user)

    def ctcpversion_command(self, args, user=None, **_unused):

        if args:
            user = args

        elif user is None:
            user = self.core.login_username

        if self.send_private(user, self.core.privatechats.CTCP_VERSION, show_ui=False):
            self.echo_message("Asked %s for client version" % user)

    def hello_command(self, args, **_unused):
        self.echo_message("Hello there! %s" % args)

    def join_chat_command(self, args, **_unused):
        self.core.chatrooms.show_room(args)

    def leave_command(self, args, room=None, **_unused):

        if args:
            room = args

        if room not in self.core.chatrooms.joined_rooms:
            self.echo_message("Not joined in room %s" % room)
            # return  # in future the gui might need to close a tab even if we are not joined, such as while offline etc

        self.core.chatrooms.remove_room(room)

    def me_chat_command(self, args, **_unused):
        self.send_message("/me " + args)

    def msg_chat_command(self, args, **_unused):

        args_split = args.split(" ", maxsplit=1)
        user, text = args_split[0], args_split[1]

        if self.send_private(user, text, show_ui=True, switch_page=False):
            self.echo_message("Private message sent to user %s" % user)

    def pm_chat_command(self, args, **_unused):
        self.core.privatechats.show_user(args)
        self.log("Private chat with user %s" % args)

    def say_chat_command(self, args, **_unused):

        args_split = args.split(" ", maxsplit=1)
        room, text = args_split[0], args_split[1]

        if self.send_public(room, text):
            self.log("Chat message sent to room %s" % room)

    def add_share_command(self, args):

        # share_type, virtual_name, path = args.split(maxsplit=3)

        self.core.shares.rescan_shares()

    def remove_share_command(self, args):

        # share_type, virtual_name, *_unused = args.split(maxsplit=2)

        self.core.shares.rescan_shares()

    def list_shares_command(self, args):
        self.echo_message("nothing here yet")

    def rescan_command(self, args, **_unused):
        self.core.shares.rescan_shares()

    def away_command(self, args, **_unused):
        self.core.set_away_mode(self.core.user_status != 1, save_state=True)  # 1 = UserStatus.AWAY
        self.echo_message("Status is now %s" % (_("Online") if self.core.user_status == 2 else _("Away")))

    def quit_command(self, args, user=None, room=None):

        if user is not None:
            interface = "private_chat"

        elif room is not None:
            interface = "chatroom"

        else:
            interface = "cli"

        if "force" not in args:
            self.log("Exiting application on %s command %s" % (interface, args))
            self.core.confirm_quit()
            return

        self.log("Quitting on %s command %s" % (interface, args))
        self.core.quit()

    def shutdown_notification(self):
        self.log("Shutdown!")
```
